docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+
+# Functionalization
+
+So far we had a pure `apply` function that is typically transformed with some JAX transformations and inside `apply` we have a stateful Module instance to work with. In other words: Outside of a Module we are in a functional world where we have the power of JAX's functional transformations and inside the Module we get the power of Flax's stateful variables and PRNG sequence, and the `apply` method is our bridge between these two worlds.
+
+But what if we want to use JAX transformations **inside** Modules? The answer to this is functionalization.
+
+This procedure itself is tedious and error-prone but handled internally by Flax. At a high-level we can summarize it as follows. For a method `fn` defined within a Module:
+
+1. Collect the state (variables & PRNG sequences) of the Module(s) that should be available inside the JAX transformation and take a snapshot of it.
+2. Call the JAX transformation with the original arguments and the collected state. Then inside the transformation:
+    1. Unpack the state and recreate the Modules
+    2. Call the user code `fn`
+    3. Collect the updated variables and rng and return it together with the original return values from `fn`
+3. Update the original state with the updated state returned from the transformation.
+
+A more in depth explanation of functionalization and lifting can be found [here](https://flax.readthedocs.io/en/latest/design_notes/lift.html).
+
+### Practical consequences
+
+For the most part functionalization is something that is handled automatically for you. Still there are some constraints that you must take into account. Most importantly, Flax only handles the stateful primitives (Linen variables and RNGs) and not arbitrary stateful Python code. Most importantly: You cannot close over stateful objects and `Module` objects because they are invisible to Flax's internals (and to JAX in general).
+
+
+```python
+class Foo(nn.Module):
+
+  @nn.compact
+  def __call__(self, x):
+    dense = nn.Dense(x.shape[-1])
+    return self.inner(x, lambda x: dense(x) + 1)
+  
+  def inner(self, x, fn):
+    for i in range(3):
+      x = fn(x)
+    return x
+```
+
+
+Here `inner` takes a function that closes over a Module instance. In this example, that works fine because we are not transforming the inner method with a lifted transformation. Most methods are not transformed but it is good to know how to make Module methods transformable.
+
+The main obstacle for transformability are types that JAX does not recognize. By default JAX will recognize containers (lists, dictionaries, and tuples) of JAX/NumPy arrays and python scalars (bool, int, float, etc.). The containers are called [pytrees](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html) and the pytree API is extensible with custom types. With the flax.struct package you can also define pytree compatible dataclasses.
```

Comment:
simpler might be something like "JAX only understands [Pytree](...) arguments. That's arbitrarily nested Python containers (dict, list, tuple) of (Jax) numpy ndarrays and Python numbers/bools. Flax allows to define dataclasses that ..."

nit: maybe link and format `flax.struct`